File: python/open_guardrail/guards/system_prompt_guard.py
"""Detects system prompt leakage in agent output."""
from __future__ import annotations

import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_DEFAULT_PATTERNS = [
    "system prompt:",
    "my instructions are",
    "i was told to",
    "my system message",
    "as an ai assistant, i am programmed",
    "my configuration is",
    "i am configured to",
    "my rules are",
    "system:",
    "<<sys>>",
    "[inst]",
    "### system",
    "role: system",
]


class _SystemPromptGuard:
    def __init__(
        self,
        *,
        action: str = "block",
        custom_patterns: Optional[List[str]] = None,
    ) -> None:
        self.name = "system-prompt-guard"
        self.action = action
        self.patterns = _DEFAULT_PATTERNS + [
            p.lower() for p in (custom_patterns or [])
        ]

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        lower = text.lower()
        matched = [p for p in self.patterns if p in lower]
        triggered = len(matched) > 0
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="system-prompt-guard",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message="System prompt leakage detected" if triggered else None,
            latency_ms=round(elapsed, 2),
            details={"matched_patterns": matched} if triggered else None,
        )


def system_prompt_guard(
    *,
    action: str = "block",
    custom_patterns: Optional[List[str]] = None,
) -> _SystemPromptGuard:
    return _SystemPromptGuard(action=action, custom_patterns=custom_patterns)

File: packages/guards/src/system-prompt-guard.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface SystemPromptGuardOptions {
  action: 'block' | 'warn';
  /** Additional custom patterns to detect */
  customPatterns?: string[];
}

const DEFAULT_PATTERNS = [
  'system prompt:',
  'my instructions are',
  'i was told to',
  'my system message',
  'as an ai assistant, i am programmed',
  'my configuration is',
  'i am configured to',
  'my rules are',
  'system:',
  '<<sys>>',
  '[inst]',
  '### system',
  'role: system',
];

export function systemPromptGuard(options: SystemPromptGuardOptions): Guard {
  const patterns = [
    ...DEFAULT_PATTERNS,
    ...(options.customPatterns ?? []).map((p) => p.toLowerCase()),
  ];

  return {
    name: 'system-prompt-guard',
    version: '0.1.0',
    description: 'Detects system prompt leakage in agent output',
    category: 'security',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const lower = text.toLowerCase();
      const matched = patterns.filter((p) => lower.includes(p));
      const triggered = matched.length > 0;
      return {
        guardName: 'system-prompt-guard',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        latencyMs: Math.round(performance.now() - start),
        details: triggered
          ? { matchedPatterns: matched }
          : undefined,
      };
    },
  };
}
